language/resource_types.md

```diff
@@ -0,0 +1,502 @@
+# Puppet Resource Types
+
+This document describes puppet's Ruby DSL for defining resource types, aka [custom types](https://docs.puppet.com/guides/custom_types.html#types). Examples are taken from the `puppetlabs-aws` module and core puppet.
+
+The second part of the document describes issues that cause puppet resource types to leak across environments (lack of environment isolation).
+
+## Puppet::Type.newtype method
+
+Defines a new puppet resource type and registers the type using the specified symbolic name. For example, to define a resource type named `ec2_instance`: 
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  # definition goes here
+end
+```
+
+The `newtype` method creates a new class `Puppet::Type::Ec2_instance`, which subclasses `Puppet::Type`. The block passed to the `newtype` method defines methods and variables on the `Puppet::Type::Ec2_instance` class.
+
+## Puppet::Type::&lt;name&gt; methods
+
+Within the `Puppet::Type.newtype` block, the following DSL methods are often called, and are evaluated in the context of the  `Puppet::Type::<name>` class:
+
+### doc string
+
+Defines the documentation string for the type, which is output when running `puppet describe <type>`. The doc string is usually set via the `Puppet::Type` class variable, though there is also a `Puppet::Type.doc=` class method.
+
+```ruby
+Puppet::Type.newtype(:ec2_instance) do
+  @doc = 'A type representing an AWS VPC'
+end
+```
+
+### namevar
+
+The type must specify a parameter to be the name of the resource, aka namevar. The name must uniquely identity that resource. By default, a parameter named `name` will be the namevar. This behavior can be modified using `title_patterns` and/or explicitly declaring a parameter to be a namevar.
```

Comment:
Is this a read-only property? Is it callable in some way? Or is it more of a concept? Since you declare this section refers to DSL methods, it's not clear how this fits in.